### DISCUSSION TITLE
HDF5 export

### DISCUSSION THREAD

**Discussion by higgx2** [danielhrisca/asammdf#688] [Q&A] [answered]
Hello,

I'm trying to export some MF4 File to HDF5 Format, but I am not able to find the group specific timestamps channel.
Can anybody explain, which part I am missing to set it up correctly?

Thx

**Comment by danielhrisca** [comment, marked as answer]
What arguments do you use when you call ``export``?

**Reply by higgx2** [reply]
only output format 'hdf5'

`mdf = MDF(mf4filename)
hdf = mdf.export('hdf5')`

I was not using the single_time_base option since I wanted to preserve the distinct time channels of each group.

I also tried to export as `mdf.export('mat', format='7.3')` but I don't like the group and channel renaming. Nevertheless a timechannel for each group would be preserved.

In both cases (hdf or mat export) I use `hdf = h5py.File(h5filename, 'r')` for reading the file.

**Reply by danielhrisca** [reply]
You will have the master channel name in the new HDF5 group as the ``master`` attribute (see here https://github.com/danielhrisca/asammdf/blob/1507f100bddc86c5f037db2770ec39e42bdfc306/asammdf/mdf.py#L1410)

**Reply by higgx2** [reply]
Correct. The name 'timestamps' is found in the attributes but the actual timechannel with the current timestamps of the group is not included or at least I am not able to find it.

I was expecting a dataset channel 'timestamps' for each group.

**Reply by danielhrisca** [reply]
I'll check this.

**Reply by higgx2** [reply]
Hi Daniel, habe you already been able to confirm my observation or am I doing something wrong?

**Reply by higgx2** [reply]
Hi Daniel (@danielhrisca),

I would like to propose to insert something like this (starting in line 1425 mdf.py):
https://github.com/danielhrisca/asammdf/blob/1507f100bddc86c5f037db2770ec39e42bdfc306/asammdf/mdf.py#L1425

                        master = self.get(group.attrs["master"], group_index)
                        if reduce_memory_usage:
                            master.timestamps = downcast(master.timestamps)
                        if compression:
                            dataset = group.create_dataset(
                                group.attrs["master"], data=master.timestamps, compression=compression
                            )
                        else:
                            dataset = group.create_dataset(
                                group.attrs["master"], data=master.timestamps, dtype=master.timestamps.dtype
                            )
                        unit = master.unit.replace("\0", "")
                        if unit:
                            dataset.attrs["unit"] = unit
                        comment = master.comment.replace("\0", "")
                        if comment:
                            dataset.attrs["comment"] = comment

**Reply by danielhrisca** [reply]
@higgx2 can you send a PR?

**Reply by higgx2** [reply]
sorry but I am failing to send a PR. I have to clone the development branch right? And then I introduce my changes commit and push, right? After that I could do a PR chosing to request a merge from dev to master? Unfortunately I am obviously not granted to send the push command...

**Reply by higgx2** [reply]
@danielhrisca:  Ok, I forked the project, insert the changes and after commiting I created a pull request from my fork (master) to the main project (master). Hopefully, I did it the right way.

**Reply by airotz** [reply]
Hi Daniel,

The export to hdf5 does not work in my case... here the traceback:

>>> mdf.export(fmt="hdf5",filename="export.hdf")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "xxx/.local/lib/python3.8/site-packages/asammdf/mdf.py", line 1463, in export
    master = self.get(group.attrs["master"], group_index)
  File "/xxx/.local/lib/python3.8/site-packages/asammdf/blocks/mdf_v4.py", line 6594, in get
    vals, timestamps, invalidation_bits, encoding = self._get_scalar(
  File "/xxx/.local/lib/python3.8/site-packages/asammdf/blocks/mdf_v4.py", line 7282, in _get_scalar
    timestamps = concatenate(timestamps, out=out)
  File "<__array_function__ internals>", line 180, in concatenate
ValueError: Output array is the wrong shape

It seems like it is having some trouble with the master... Could you please confirm?

Thank you very much in advance for your answer.

Best regards,
Igor
